Make category pages indexable

bobsilverberg opened the pull request:
Fixes #10069

willdurand commented on src/amo/components/Routes/index.js:
```diff
@@ -147,6 +147,12 @@ const Routes = ({ _config = config }: Props = {}) => (
 
+    <Route
+      exact
+      path="/:lang/:application/category/"
```
We probably want per-category URLs so it should be something like this:

```
path="/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/"
```

bobsilverberg replied:
Why do we need those? I thought this was just about the ability to crawl. I can change it, but this is the most expedient way to accomplish this.

willdurand replied:
Pretty urls are a bit better in terms of crawling experience and we're going to enhance these pages in the future. I think it's best to plan for the future, it's not a lot of work.

We can mimic what is done in `SearchPage`. We don't need to worry about the server redirects, etc. so the `CategoryPage` should be a simple component (famous last words?)

willdurand commented on src/amo/components/Routes/index.js:
```diff
@@ -147,6 +147,12 @@ const Routes = ({ _config = config }: Props = {}) => (
+    <Route
+      exact
+      path="/:lang/:application/category/"
+      component={(props) => <SearchPage {...props} isForCategory />}
```
It'd be best to create a new page component (`CategoryPage`?). We'll want to use `HeadLinks` and maybe `HeadMetaTags` in the future.

bobsilverberg replied:
Again, I'm not sure if we need these if the only point is to have a page for the crawler to crawl. Will the page be penalized if it doesn't contain those?

bobsilverberg replied:
FWIW, I did start with a new page component and then realized "hey, we don't need that", but I guess maybe we do.

willdurand replied:
Yeah, I think we do need a new page component. We need `HeadLinks` to tell the crawlers what to crawl.

bobsilverberg commented:
We also need to have ops add a route to nginx for this new route. Is that generally done via a bug or via an issue?

bobsilverberg commented on src/amo/components/Search/index.js:
```diff
@@ -115,100 +116,104 @@ export class SearchBase extends React.Component<InternalProps> {
-    let title = i18n.gettext('Search results');
+    let title = pageTitle;
 
-    if (filters.promoted) {
```
FYI, the way GitHub has interpreted this diff, makes it looks like there are a lot of changes, but in fact the code below is completely unchanged. The only change is to introduce the use of the `pageTitle` prop as an override.

bobsilverberg commented:
@willdurand This is ready for another look.

willdurand commented:
> We also need to have ops add a route to nginx for this new route. Is that generally done via a bug or via an issue?

do we?

bobsilverberg commented:
> > We also need to have ops add a route to nginx for this new route. Is that generally done via a bug or via an issue?
> 
> do we?

I recall having to do it before, and also, there's this comment: https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Routes/index.js#L42, but maybe I'm wrong?

@bqbn Do we need anything in nginx to support a new route in our app? We've added a route for `path="/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/"`.

willdurand commented:
> I recall having to do it before, and also, there's this comment: [https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Routes/index.js#L42](https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Routes/index.js?rgh-link-date=2021-03-04T19%3A34%3A21Z#L42), but maybe I'm wrong?

I was looking at this comment. Given that the sentence doesn't really seem written in English, I guess I wrote it? 😅

bobsilverberg commented:
> > I recall having to do it before, and also, there's this comment: [https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Routes/index.js#L42](https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Routes/index.js?rgh-link-date=2021-03-04T19%3A36%3A00Z#L42), but maybe I'm wrong?
> 
> I was looking at this comment. Given that the sentence doesn't really seem written in English, I guess I wrote it? 😅

Blame agrees 😉

bqbn commented:
> > > We also need to have ops add a route to nginx for this new route. Is that generally done via a bug or via an issue?
> > 
> > 
> > do we?
> 
> I recall having to do it before, and also, there's this comment: https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Routes/index.js#L42, but maybe I'm wrong?
> 
> @bqbn Do we need anything in nginx to support a new route in our app? We've added a route for `path="/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/"`.

Please take a look at [this comment](https://github.com/mozilla-services/cloudops-deployment/blob/dc76521ac40e8c5b29c0d50208c98a793c2dbda4/projects/amo/puppet/modules/amo_proxy/templates/nginx.addons.conf.erb#L352). Based on the 4 tickets mentioned in that comment, there are 4 existing redirection rules that deal with the `/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/` path. Please make sure that doesn't conflict with what you're trying to achieve here.

That said, by default, Nginx routes everything under `/` to the frontend.

willdurand commented:
> Please take a look at [this comment](https://github.com/mozilla-services/cloudops-deployment/blob/dc76521ac40e8c5b29c0d50208c98a793c2dbda4/projects/amo/puppet/modules/amo_proxy/templates/nginx.addons.conf.erb#L352). Based on the 4 tickets mentioned in that comment, there are 4 existing redirection rules that deal with the `/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/` path. Please make sure that doesn't conflict with what you're trying to achieve here.

ugh, that doesn't sound good.

bobsilverberg commented:
@jvillalobos Because of some permanent redirects put in place back in May or 2019, we cannot reuse the old URL path for category pages, which was `/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/`. We need some alternate variation. A few suggestions are:

- `/:lang/:application/:visibleAddonType(extensions|themes)/category/:categorySlug/`
- `/:lang/:application/:visibleAddonType(extensions|themes)/c/:categorySlug/`
- `/:lang/:application/c/:visibleAddonType(extensions|themes)/:categorySlug/`

What do you think/recommend?

bqbn commented:
Why can't we remove [the 4 redirection rules](https://github.com/mozilla-services/cloudops-deployment/blob/dc76521ac40e8c5b29c0d50208c98a793c2dbda4/projects/amo/puppet/modules/amo_proxy/templates/nginx.addons.conf.erb#L353-L366) from Nginx config?

bobsilverberg commented:
> Why can't we remove [the 4 redirection rules](https://github.com/mozilla-services/cloudops-deployment/blob/dc76521ac40e8c5b29c0d50208c98a793c2dbda4/projects/amo/puppet/modules/amo_proxy/templates/nginx.addons.conf.erb#L353-L366) from Nginx config?

If we remove the redirects then when someone does visit one of those routes it will be a 404, because the new route for the category page will not follow the pattern of `/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/`, unless I am mistaken.

jvillalobos commented:
> /:lang/:application/:visibleAddonType(extensions|themes)/category/:categorySlug/

That one looks fine, but I'm also not clear about the redirects.

> the new route for the category page will not follow the pattern of /:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/

Aren't we essentially trying to bring back the old URLs? Why wouldn't it work to drop the redirects?

willdurand commented:
> Aren't we essentially trying to bring back the old URLs? Why wouldn't it work to drop the redirects?

We can't, that would lead to redirect loops in some cases because existing redirects are permanent. This is why we need a new URL pattern.

willdurand commented on src/amo/pages/CategoryPage/index.js:
```diff
@@ -0,0 +1,143 @@
+
+    // The only two possibilites for addonType are 'extension' and
+    // 'statictheme' because anything else would have thrown an
+    // exception in mapStateToProps via apiAddonType()
```
The router should reject invalid values because the `path` is a regular expression.

bobsilverberg replied:
That's true, but the result is the same. This comment is explaining why we don't need to address any different add-on types. I guess I can just remove the comment, or would you prefer that I update it to explain your reason why we don't need to worry about other cases?

willdurand commented on src/amo/pages/CategoryPage/index.js:
```diff
@@ -0,0 +1,143 @@
+    // The only two possibilites for addonType are 'extension' and
+    // 'statictheme' because anything else would have thrown an
+    // exception in mapStateToProps via apiAddonType()
+    if (addonType === ADDON_TYPE_EXTENSION) {
```
I'd rather special case themes because we could _potentially_ see a theme as an extension but the opposite isn't true.

willdurand commented on src/amo/pages/CategoryPage/index.js:
```diff
@@ -0,0 +1,143 @@
+  render() {
+    const { categoryName, filters, match } = this.props;
+    const { categorySlug, visibleAddonType } = match.params;
+    const addonType = apiAddonType(visibleAddonType);
```
This could be injected in `mapStateToProps`.

bobsilverberg replied:
I started by doing that and it was creating an issue/complication, but I don't remember what it was. I can try again.

willdurand commented on src/amo/pages/CategoryPage/index.js:
```diff
@@ -0,0 +1,143 @@
+  }
+}
+
+function mapStateToProps(state: AppState, ownProps: Props): PropsFromState {
```
Although the codebase is still inconsistent, we tend to prefer arrow functions.

willdurand commented on src/amo/pages/CategoryPage/index.js:
```diff
@@ -0,0 +1,143 @@
+
+      if (categories && categories[categorySlug]) {
+        categoryName = categories[categorySlug].name;
+      }
```
This logic could be moved into a helper function.

bobsilverberg replied:
It looks like it's only used once, here, so what's the advantage? Just to make the `mapStateToProps` function smaller/cleaner?

bobsilverberg replied:
Oh, wait, it's also in `SearchContextCard`, just using different variable names. 😛

willdurand commented on tests/unit/amo/pages/TestSearchPage.js:
```diff
@@ -153,6 +157,28 @@ describe(__filename, () => {
     sinon.assert.notCalled(fakeDispatch);
   });
 
+  it('dispatches a server redirect when `category` and `addonType` is set', () => {
```
why is it "is set" and not "are set"?

bobsilverberg replied:
Because my English is poor (or I wasn't paying attention).

willdurand commented on tests/unit/amo/pages/TestSearchPage.js:
```diff
@@ -153,6 +157,28 @@ describe(__filename, () => {
+      fakeDispatch,
+      sendServerRedirect({
+        status: 301,
+        url: `/en-US/android/extensions/category/${category}/?page=${page}`,
```
Maybe explicitly pass the `lang` or `clientApp` in this test case to make sure the values are updated correctly.

willdurand reviewed (commented):
Some comments while I am testing the feature locally

willdurand commented on src/amo/pages/SearchTools/index.js:
```diff
@@ -3,85 +3,50 @@ import * as React from 'react';
         status: 301,
-        url: `/${lang}/${clientApp}/search/${queryString}`,
+        url: `/${lang}/${clientApp}${pathname}${makeQueryString({
+          sort: SEARCH_SORT_TOP_RATED,
```
mm it isn't the same thing as before, right? I don't see a `sort` filter in the previous code?

bobsilverberg replied:
It was in `mapStateToProps` in the old code, but looking at the current behaviour of the site it isn't being picked up, so I guess we can remove it.

bobsilverberg commented on src/amo/components/SearchFilters/index.js:
```diff
@@ -105,15 +105,24 @@ export class SearchFiltersBase extends React.Component<InternalProps> {
+
+    // If category is a filter, remove category and type as they are already
+    // included in pathname.
+    if (filters.category) {
```
Good point. It shouldn't ever be needed.

bobsilverberg replied:
I think this might be the comment that made me remove the code that removes the category and type filters, but I think we did in fact need that code, as it was used to remove them when a user changes the sort sequence from the SearchFilters component. I'll add it back in and see if that fixes it.

bobsilverberg commented:
Note that not all of the comments have been addressed yet, but I am eoding, so I'm going to push a commit that addresses most of them.

bobsilverberg commented:
I have now addressed all of the comments. I did find one regression when testing though, and I don't have time to address it this afternoon, but I should be able to deal with it next week.

If you are on the new CategoryPage and you type a query into the Search input in the header, it no longer searches within the category. I believe this is because the category is no longer in the URL as a query param, but I am guessing there is another way for us to address this, other than adding it back in. I'll look into it next week.

willdurand commented:
> If you are on the new CategoryPage and you type a query into the Search input in the header, it no longer searches within the category. I believe this is because the category is no longer in the URL as a query param, but I am guessing there is another way for us to address this, other than adding it back in. I'll look into it next week.

Given that we are not on a "search results" page anymore, do we want to keep this behavior?

bobsilverberg commented:
> > If you are on the new CategoryPage and you type a query into the Search input in the header, it no longer searches within the category. I believe this is because the category is no longer in the URL as a query param, but I am guessing there is another way for us to address this, other than adding it back in. I'll look into it next week.
> 
> Given that we are not on a "search results" page anymore, do we want to keep this behavior?

Hmm, maybe not? What do you think @jvillalobos ?

jvillalobos commented:
Honestly I didn't remember we did this 😀. It's a nice feature, but not really discoverable. I think it should be fine to remove it.

bobsilverberg commented:
@willdurand I'd really like to get this into this week's tag, so if you have time to take another look today, or tomorrow morning, that would be great.

willdurand commented:
> @willdurand I'd really like to get this into this week's tag, so if you have time to take another look today, or tomorrow morning, that would be great.

So this is ready for another look? No more regression to fix?

bobsilverberg commented:
> > @willdurand I'd really like to get this into this week's tag, so if you have time to take another look today, or tomorrow morning, that would be great.
> 
> So this is ready for another look? No more regression to fix?

Correct, that's why I asked. ;-)  Jorge seems fine with the one "regression" listed above. I didn't find any others.

willdurand commented on src/amo/pages/CategoryPage/index.js:
```diff
@@ -0,0 +1,146 @@
+const CategoryPage: React.ComponentType<Props> = compose(
+  connect(mapStateToProps),
+  translate(),
+  withErrorHandler({ name: 'CategoryPage' }),
```
Shouldn't we be using a "fixed" error handler given that it's a page-level component?

willdurand commented on src/amo/components/SearchFilters/index.js:
```diff
@@ -114,15 +114,17 @@ export class SearchFiltersBase extends React.Component<InternalProps> {
   doSearch(newFilters: SearchFiltersType) {
     const { clientApp, lang, history, pathname } = this.props;
 
-    if (newFilters.page) {
```
Why is this change needed actually?

bobsilverberg replied:
It's not, anymore. Subsequent changes have rendered it unnecessary. Thanks for pointing it out.

bobsilverberg replied:
And... this code is back now 😉

willdurand commented on src/amo/pages/SearchPage/index.js:
```diff
@@ -43,9 +45,23 @@ export class SearchPageBase extends React.Component<InternalProps> {
+      });
+
+      delete newFilters.addonType;
+      delete newFilters.category;
```
I think we should also delete `sort` if it's equal to the default sort because we should redirect to the canonical category url when possible.

willdurand commented on src/amo/pages/CategoryPage/index.js:
```diff
@@ -0,0 +1,146 @@
+  withErrorHandler({ name: 'CategoryPage' }),
+)(CategoryPageBase);
+
+export default CategoryPage;
```
Shouldn't we return a 404 when the category slug is invalid? For example: http://localhost:3000/en-US/firefox/themes/category/nope/

bobsilverberg replied:
Good catch. I assumed the logic in `Page` would take care of that, but the API error is coming from the `Search` component. I think the best way to address this is to just change the logic of the `Search` component to display a 404 page if the API response is either `404` or `400`. I can't think of an occasion on which we'd want to display anything other than a 404 page when the search API returns a 400. I've made this change and you can see what you think when I push the commit.

willdurand replied:
I think it's fine (and it will be even better with a comment) but we should probably file an issue because for other search parameters, we display the error on the search page, which is kinda nice too.

willdurand reviewed (commented):
I have a few more comments.

bobsilverberg commented:
Hmm, I'm seeing a new/old issue. When you change the sort order from the SearchFilters component on the CategoryPage, it adds the `category` and `type` to the URL. I know I had code to remove those before, but now they're back. I think maybe there was a comment directing me to remove the code I had for that, but now I think we need it again.

bobsilverberg commented:
I fixed the issue with the category and type being added to the URL by bringing back the code I wrote earlier for `SearchFilters`.

bobsilverberg commented:
@willdurand This is ready for another look.

willdurand commented on src/amo/components/Search/index.js:
```diff
@@ -251,7 +256,7 @@ export class SearchBase extends React.Component<InternalProps> {
       log.warn(`Captured API Error: ${errorHandler.capturedError.messages}`);
 
-      if (errorHandler.capturedError.responseStatusCode === 404) {
+      if ([400, 404].includes(errorHandler.capturedError.responseStatusCode)) {
```
We should document that because it might not be obvious in the future. I guess that's because the API rejects the invalid category?

bobsilverberg replied:
Yes, it does. It sends us a specific message about "invalid category", but I figured it made more sense to just 404 on any 400 than try to look for that specific message. Where would you recommend we document this? Just in a comment above this?

willdurand replied:
the comment is perfect.

willdurand reviewed (approved):
Looks good to me, thanks!

I noticed one other thing but maybe it's fine for now: when we go to a category page, change a filter on the side, change back to how it was initially, the query string remains in the URL.

r+wc

bobsilverberg commented:
> Looks good to me, thanks!
> 
> I noticed one other thing but maybe it's fine for now: when we go to a category page, change a filter on the side, change back to how it was initially, the query string remains in the URL.
> 
> r+wc

I'm not too concerned about that. If we really think it's an issue we can file a follow-up, as it would be nice to land this for today's tag.
